[SQL] AS OF JOIN on ROW columns produces error[E0609]

Following SQL:

```
CREATE TABLE tbl1(
id INT,
roww ROW(i1 INT, v1 VARCHAR NULL));

CREATE TABLE tbl2(
id INT,
roww ROW(i1 INT, v1 VARCHAR NULL));

CREATE MATERIALIZED VIEW v AS SELECT
t1.id, t1.roww AS t1_roww, t2.roww AS t2_roww
FROM tbl1 t1
LEFT ASOF JOIN tbl2 t2
MATCH_CONDITION ( t1.roww >= t2.roww)
ON t1.id = t2.id;
```
Given example input: 

1. **data1** for **tbl1**

```
data1 = [
    {"id": 1, "roww": {"i1": 12, "v1": None}},
    {"id": 2, "roww": {"i1": 2, "v1": None}}]
```

2. **data2** for **tbl2**

```
data2 = [
    {"id": 1, "roww": {"i1": 13, "v1": None}},
    {"id": 2, "roww": {"i1": 1, "v1": None}}]
```


Produces the following error:

> **error[E0609]: no field `0` on type `&std::option::Option<dbsp::utils::Tup2<i32, std::option::Option<feldera_sqllib::SqlString>>>`
>   --> crates/feldera_pipe_operator_c5d24357095afcaa/src/lib.rs:74:38
>    |
> 74 |             let ord = compare_(&left.0, &right.0, true, true);
>    |                                      ^ unknown field
>  error[E0609]: no field `0` on type `&std::option::Option<dbsp::utils::Tup2<i32, std::option::Option<feldera_sqllib::SqlString>>>`
>   --> crates/feldera_pipe_operator_c5d24357095afcaa/src/lib.rs:74:48
>    |
> 74 |             let ord = compare_(&left.0, &right.0, true, true);
>    |                                                ^ unknown field
>  error[E0609]: no field `1` on type `&std::option::Option<dbsp::utils::Tup2<i32, std::option::Option<feldera_sqllib::SqlString>>>`
>   --> crates/feldera_pipe_operator_c5d24357095afcaa/src/lib.rs:76:38
>    |
> 76 |             let ord = compare_(&left.1, &right.1, true, true);
>    |                                      ^ unknown field
>  error[E0609]: no field `1` on type `&std::option::Option<dbsp::utils::Tup2<i32, std::option::Option<feldera_sqllib::SqlString>>>`
>   --> crates/feldera_pipe_operator_c5d24357095afcaa/src/lib.rs:76:48
>    |
> 76 |             let ord = compare_(&left.1, &right.1, true, true);
>    |                                                ^ unknown field**

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/CalciteToDBSPCompiler.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/CalciteToDBSPCompiler.java
@@ -1883,8 +1883,8 @@ public class CalciteToDBSPCompiler extends RelVisitor
 
     private void visitAsofJoin(LogicalAsofJoin join) {
         // This shares a lot of code with the LogicalJoin
-        IntermediateRel node = CalciteObject.create(join);
         CalciteObject conditionNode = CalciteObject.create(join, join.getCondition());
+        IntermediateRel node = CalciteObject.create(join, conditionNode.getPositionRange());
         JoinRelType joinType = join.getJoinType();
         boolean isLeft = joinType == JoinRelType.LEFT_ASOF;
         if (!isLeft)
@@ -2049,17 +2049,20 @@ public class CalciteToDBSPCompiler extends RelVisitor
 
         DBSPVariablePath leftVar = leftElementType.ref().var();
         DBSPExpression leftTS = leftVar.deref().field(leftTsIndex);
+        CalciteObject matchNode = CalciteObject.create(join, join.getMatchCondition());
         DBSPType leftTSType = leftTS.getType();
+        if (leftTSType.is(DBSPTypeTupleBase.class) || leftTSType.is(DBSPTypeStruct.class))
+            throw new UnimplementedException("Join on struct types", 3398, matchNode);
 
         // Currently not used
-        DBSPComparatorExpression comparator = new DBSPNoComparatorExpression(node, leftTSType);
+        DBSPComparatorExpression comparator = new DBSPNoComparatorExpression(matchNode, leftTSType);
         boolean ascending = comparison == SqlKind.GREATER_THAN || comparison == SqlKind.GREATER_THAN_OR_EQUAL;
         if (comparison != SqlKind.GREATER_THAN_OR_EQUAL) {
             // Not yet supported by DBSP
             throw new UnimplementedException(
                     "Currently the only MATCH_CONDITION comparison supported by ASOF joins is '>='", 2212, node);
         }
-        comparator = new DBSPDirectComparatorExpression(node, comparator, ascending);
+        comparator = new DBSPDirectComparatorExpression(matchNode, comparator, ascending);
 
         DBSPVariablePath k = keyType.ref().var(conditionNode);
         DBSPVariablePath l0 = leftElementType.ref().var();

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/monotonicity/Monotonicity.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/monotonicity/Monotonicity.java
@@ -482,12 +482,13 @@ public class Monotonicity extends CircuitVisitor {
 
     @Override
     public void postorder(DBSPMapOperator node) {
-        MonotoneExpression inputFunction = this.getMonotoneExpression(node.input());
+        OutputPort input = node.input();
+        MonotoneExpression inputFunction = this.getMonotoneExpression(input);
         if (inputFunction == null)
             return;
         MonotoneTransferFunctions mm = new MonotoneTransferFunctions(
                 this.compiler(), node,
-                MonotoneTransferFunctions.ArgumentKind.fromType(node.input().outputType()),
+                MonotoneTransferFunctions.ArgumentKind.fromType(input.outputType()),
                 getBodyType(inputFunction));
         MonotoneExpression result = mm.applyAnalysis(node.getFunction());
         if (result == null)
@@ -664,7 +665,13 @@ public class Monotonicity extends CircuitVisitor {
 
         DBSPExpression leftTsField = l.deepCopy().deref().field(node.leftTimestampIndex);
         DBSPExpression rightTsField = r.deepCopy().deref().field(node.rightTimestampIndex);
-        DBSPExpression min = ExpressionCompiler.makeBinaryExpression(node.getNode(), rightTsField.getType(),
+        DBSPType rightTsType = rightTsField.getType();
+        if (rightTsType.mayBeNull) {
+            // The function takes non-nullable timestamps
+            rightTsField = rightTsField.nullabilityCast(rightTsType.withMayBeNull(false), false);
+        }
+        // We know the right timestamp is not nullable, so the result type is from the left timestamp
+        DBSPExpression min = ExpressionCompiler.makeBinaryExpression(node.getNode(), leftTsField.getType(),
                 DBSPOpcode.MIN, leftTsField, rightTsField);
 
         List<DBSPExpression> fields = new ArrayList<>();

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/IncrementalRegressionTests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/IncrementalRegressionTests.java
@@ -1553,4 +1553,61 @@ public class IncrementalRegressionTests extends SqlIoTest {
                                                 new DBSPI32Literal(2, true),
                                                 new DBSPI32Literal(3, true)))))));
     }
+
+    @Test
+    public void asofTest1() {
+        this.getCC("""
+                CREATE TABLE S (
+                   id uuid,
+                   ts TIMESTAMP LATENESS INTERVAL 1 MONTH
+                );
+                
+                CREATE TABLE C (
+                    id uuid,
+                    ts timestamp NOT NULL LATENESS INTERVAL 1 HOUR
+                );
+                
+                create view V0 AS
+                SELECT c.*
+                FROM C LEFT ASOF JOIN S
+                    MATCH_CONDITION ( c.ts >= s.ts )
+                    ON c.id = s.id;
+                
+                CREATE VIEW V1 AS
+                SELECT ts FROM V0;""");
+    }
+
+    @Test
+    public void issue5032a() {
+        this.statementsFailingInCompilation("""
+                CREATE TABLE tbl1(id INT,
+                roww ROW(i1 INT, v1 VARCHAR NULL) NOT NULL);
+                
+                CREATE TABLE tbl2(id INT,
+                roww ROW(i1 INT, v1 VARCHAR NULL) NOT NULL);
+                
+                CREATE MATERIALIZED VIEW v AS SELECT
+                t1.id, t1.roww AS t1_roww, t2.roww AS t2_roww
+                FROM tbl1 t1
+                LEFT ASOF JOIN tbl2 t2
+                MATCH_CONDITION ( t1.roww >= t2.roww)
+                ON t1.id = t2.id""", "Not yet implemented: Join on struct types");
+    }
+
+    @Test
+    public void issue5032() {
+        this.statementsFailingInCompilation("""
+                CREATE TABLE tbl1(id INT,
+                roww ROW(i1 INT, v1 VARCHAR NULL));
+                
+                CREATE TABLE tbl2(id INT,
+                roww ROW(i1 INT, v1 VARCHAR NULL));
+                
+                CREATE MATERIALIZED VIEW v AS SELECT
+                t1.id, t1.roww AS t1_roww, t2.roww AS t2_roww
+                FROM tbl1 t1
+                LEFT ASOF JOIN tbl2 t2
+                MATCH_CONDITION ( t1.roww >= t2.roww)
+                ON t1.id = t2.id;""", "Not yet implemented: Join on struct types");
+    }
 }

